File: prac/segment.py
#segment tree
class node:
    def __init__(self,r):
        self.r=r
        self.left=None
        self.right=None
        self.val=False


def maketree(arr,root):
    if root.r[0]!=root.r[1]:
        #print(root.r[0],root.r[1])
        root.left=node((root.r[0],(root.r[0]+root.r[1])//2))
        root.right=node(((root.r[0]+root.r[1])//2+1,root.r[1]))
        maketree(arr,root.left)
        maketree(arr,root.right)
    elif(root.r[0]==root.r[1]):
        root.val=arr[root.r[0]]
        #print(root.r[0],root.r[1],root.val)

def enum(root):
    if root.val:
        a=0
    elif root.left.val and root.right.val:
        #print(root.r[0],root.r[1],root.val,root.left.val, root.right.val)
        root.val=root.left.val+root.right.val
    elif not root.val:
        #print(root.r[0],root.r[1],root.val,root.left.val, root.right.val)
        enum(root.left)
        enum(root.right)
        

def inorder(root):
    if root:
        inorder(root.left)
        print((root.r),root.val)
        inorder(root.right)

def query(interval,root):
    l=interval[0]
    h=interval[1]
    if l<=root.r[0] and h>=root.r[1]:
        return root.val
    elif (l>root.r[1] and h>root.r[1]) or (l<root.r[0] and h<root.r[0]):
        return 0
    else:
        return query(interval,root.left) + query(interval,root.right)

    

#arr=[1,3,5,7,9,11]
arr=[]
n=int(input("enter length of array : "))
for i in range(0,n):
    e=int(input())
    arr.append(e)

root=node((0,len(arr)-1))
maketree(arr,root)
while not root.val:
    enum(root)

print("inorder : ")
inorder(root)
print("queries")
while True:
    l=int(input("enter lower : "))
    h=int(input("enter high :"))
    print(query((l,h),root))
    c=int(input("0==exit 1==continue"))
    if c==0:
        break

File: prac/segment.java
class node{
    int low;
    int high;
    int val;
    node right;
    node left;
}
public class segment
{
    static node root=new node();
    public static void inorder(node root){
        if (root!=null){
            inorder(root.left);
            //System.out.print("("+root.low+","+root.high+")");
            System.out.print(root.val+" ");
            inorder(root.right);
        }
    }
    public static void build(node root,int arr[]){
        node temp=root;
        if (temp.low!=temp.high){
            temp.left=new node();
            temp.left.val=0;
            temp.right=new node();
            temp.right.val=0;
            temp.left.low=temp.low;
            temp.left.high=(int)(temp.low+temp.high)/(int)(2);
            temp.right.low=1+(int)(temp.low+temp.high)/(int)(2);
            temp.right.high=temp.high;
            build(temp.left,arr);
            build(temp.right,arr);
        }
        else{
            temp.val=arr[temp.low];
        }
    }
    public static void enumerate(node root){
        if (root.val!=0){
            int tp=0;
        }
        else if (root.left.val!=0 && root.right.val!=0){
            root.val=root.left.val+root.right.val;
            System.out.println(root.val+ " ("+root.low+" "+root.high+")");
        }
        else{
            if (root.left!=null)
            enumerate(root.left);
            if (root.right!=null)
            enumerate(root.right);
        }
    }
    public static int query(int l,int h,node root){
        
        if (root.low>=l && root.high <=h){
            //System.out.println(root.val);
            return root.val;
        }
        else if(root.low==root.high && (root.low<l || root.high >h))
        return 0;
        else
         return query(l,h,root.left)+query(l,h,root.right);
    }
	public static void main(String[] args) {
	    int arr[]={1,2,3,4,5,6,7};
	    root.low=0;
	    root.high=arr.length-1;
	    build(root,arr);
	    //System.out.println(root.val);
	    while(root.val==0){
	        enumerate(root);
	    }
	    inorder(root);
	    int ans=query(1,5,root);
	    System.out.println("\n ans : "+ans);
	    
		//System.out.println(arr);
	}
}
